### PR TITLE
Feature auth

### PR DESCRIPTION
**Pull request by MaHeak**
Adding feature auth into v 1.0

### CODE REVIEW

File: backend/src/main.js [hunk 1]
```diff
@@ -1,114 +1,178 @@
...
-    });
+    })(req, res, next);
 });
 
-app.get('/api/entries/:id', function (req, res, next) {
-    passEntry.find({'id': res.params[0]}, function(err, passentry){
-        if (err){
+app.get('/api/entries/:id', async function (req, res, next) {
+    try {
+        const passEntryOne = await passEntry.findOne({ where: {ID: res.params[0]} });
```
Comment (Eoksni) [inline backend/src/main.js]: It seems its is not protected by passport.

File: backend/src/main.js [hunk 2]
```diff
@@ -1,114 +1,178 @@
...
+            });
+        } else {
+            res.sendStatus(401);
+        }
+    } else {
+        res.sendStatus(401);
+    }
+});*/
+
+app.delete('/api/token',  function (req, res, next) {
```
Comment (Eoksni) [inline backend/src/main.js]: Token is created by using /api/login and deleted by using /api/token. According to REST, we need to use same url for all actions for the same resource. So stick to one of the /api/login or /api/token in both cases.

Comment (Eoksni) [inline backend/src/main.js]: And it should be protected from unauthorized access.

File: backend/src/main.js [hunk 3]
```diff
@@ -1,114 +1,178 @@
...
-            log.info("new password entry created");
-            res.status = 201;
-            res.location(`/api/entries/${entryid}`);
+    return passport.authenticate('jwt', {session: 'false'}, async function (err, user, info){
+        try {
+            let newentryID = await passEntry.create({
+            userID: user.ID,
+            title: req.body.title,
+            user: req.body.user,
+            password: req.body.password}).get('ID');
```
Comment (Eoksni) [inline backend/src/main.js]: why did you rename `id` to `ID`?

File: backend/src/main.js [hunk 4]
```diff
@@ -1,114 +1,178 @@
...
-app.put('/api/entries/:id', function (req, res) {
+app.put('/api/entries/:id', async function (req, res) {
     const updatedEntry = req.body;
-    updatedEntry.id = req.params[0];
-    if (updatedEntry.title === undefined || updatedEntry.title === "") {
-        res.status = 400;
-        res.body = { reason: "request must specify non-empty title" };
-        return;
+    try{
+        await passEntry.update({title: req.body.title, user: req.body.user, password: req.body.password}, {where: {ID: req.params[0]}});
```
Comment (Eoksni) [inline backend/src/main.js]: unauthorized access

File: backend/src/main.js [hunk 5]
```diff
@@ -1,114 +1,178 @@
...
-app.delete('/api/entries/:id', function (req, res) {
-    passEntry.deleteOne(req.params[0], function(err){
-        if (err){
-            res.statusCode = 500;
-            log.error('Internal error(%d): %s',res.statusCode,err.message);
-            return res.send({ error: 'Server error' });
-        }
+app.delete('/api/entries/:id', async function (req, res) {
+    try{
+        await passEntry.destroy({where: {ID: req.params[0]}});
```
Comment (Eoksni) [inline backend/src/main.js]: unauthorized access

File: backend/src/models/user.js [hunk 6]
```diff
@@ -1,35 +1,56 @@
...
+    passwordHash: {
+        type: Sequelize.TEXT,
+        allowNull: false
+    },
+    salt: {
+        type: Sequelize.STRING,
+        allowNull: false
+    },
+    token: {
+        type: Sequelize.STRING
```
Comment (Eoksni) [inline backend/src/models/user.js]: why do you keep token in the db model? what if user has several devices and several tokens for each of them ?

File: package.json [hunk 7]
```diff
@@ -5,18 +5,27 @@
...
+    "pg-hstore": "^2.3.2",
+    "sequelize": "^4.3.2",
     "toastr": "^2.1.2",
     "vue": "^2.2.4",
     "vue-bootstrap-modal": "github:eoksni/vue-bootstrap-modal#905ce6b16992b238e8b9394976e9b2b049400cbb",
     "vue-i18n": "^5.0.3",
     "vue-resource": "^1.2.1",
     "vue-router": "^2.3.0",
-    "vuex": "^2.2.1"
+    "vuex": "^2.2.1",
```
Comment (Eoksni) [inline package.json]: what do you need vuex for?

Comment (Eoksni) [inline package.json]: oh, it is actually a leftover, nvm

File: backend/src/main.js (hunk 1 above)
Comment (MaHeak) [inline backend/src/main.js]: Yeah, because this ought to be working like this: unregistered users can add and edit password entries, but list of entries different for all users. When you enter the page and login express give you your list entries and you can edit only entries, who connected with your userID in db.

File: backend/src/main.js (hunk 2 above)
Comment (MaHeak) [inline backend/src/main.js]: I think a lot about it, and when i look in the OAuth 2.0, i see that login and token are two different APIs for different functionality. But in our situation we have only two actions and they must be according to REST. But i think, that is so... strange may be, that i must login on /api/token.

File: backend/src/main.js (hunk 1 above)
Comment (Eoksni) [inline backend/src/main.js]: But won't it be the same list for all unregistered users?

File: backend/src/main.js (hunk 3 above)
Comment (MaHeak) [inline backend/src/main.js]: It's a habit from cache db, i'll try to name fields in db like sNote or oUser ("s" means string, "o" means object).

File: backend/src/main.js (hunk 1 above)
Comment (MaHeak) [inline backend/src/main.js]: I have something TODO for this situation. Unregistered users enters the page and create records with userID null, but IDs for this records stored in local client storage. And when they reenters the page, backend will get records from base by IDs from users local storage.

File: backend/src/main.js (hunk 4 above)
Comment (MaHeak) [inline backend/src/main.js]: Same situation like at https://github.com/emotz/passwordkeeper/pull/60/files#r127956868

File: backend/src/models/user.js (hunk 6 above)
Comment (MaHeak) [inline backend/src/models/user.js]: It's just a field, i don't store a token into db

File: backend/src/main.js (hunk 1 above)
Comment (Eoksni) [inline backend/src/main.js]: it is still very not right to be able to get any passentry just knowing its id. User's passentry should *only* be accessible by people who have corresponding access token.

Comment (MaHeak) [inline backend/src/main.js]: Registered users get there recirds AFTER corresponding access token

File: backend/src/main.js (hunk 2 above)
Comment (Eoksni) [inline backend/src/main.js]: I agree it seems kinda weird, but still, we should stick to REST. And in REST every api entry should correspond to a resource. You can't really say that `login` is a resource. But you can easily say that `token` is a resource.

File: backend/src/main.js (hunk 3 above)
Comment (Eoksni) [inline backend/src/main.js]: This is too late for that kind of change, at least not in current circumstances - renaming in js is hard, and we don't have much spare time. Just stick to the naming that we had.

File: backend/src/models/user.js (hunk 6 above)
Comment (Eoksni) [inline backend/src/models/user.js]: Why do you have a field for it then?

File: backend/src/main.js (hunk 1 above)
Comment (Eoksni) [inline backend/src/main.js]: yes, but unregistered user can easily get any record of any user just by knowing or guessing record id, and not only get, but change it as well. that is pretty BS :)

**Comment by MaHeak** [comment]
In last commen i add some changes
